utils/chunked_vector.hh

```diff
@@ -430,18 +432,33 @@ chunked_vector<T, max_contiguous_allocation>::make_room(size_t n, bool stop_afte
 }
 
 template <typename T, size_t max_contiguous_allocation>
-void
-chunked_vector<T, max_contiguous_allocation>::do_reserve_for_push_back() {
+typename std::pair<typename chunked_vector<T, max_contiguous_allocation>::chunk_ptr, size_t> chunked_vector<T, max_contiguous_allocation>::get_chunk_before_emplace_back() {
+    size_t new_chunk_capacity;
     if (_capacity == 0) {
         // allocate a bit of room in case utilization will be low
-        reserve(std::clamp(512 / sizeof(T), size_t(1), max_chunk_capacity()));
+        new_chunk_capacity = std::clamp(512 / sizeof(T), size_t(1), max_chunk_capacity());
     } else if (_capacity < max_chunk_capacity() / 2) {
         // exponential increase when only one chunk to reduce copying
-        reserve(_capacity * 2);
+        new_chunk_capacity = _capacity * 2;
     } else {
         // add a chunk at a time later, since no copying will take place
-        reserve((_capacity / max_chunk_capacity() + 1) * max_chunk_capacity());
+        new_chunk_capacity = max_chunk_capacity();
+    }
+    return std::make_pair(new_chunk(new_chunk_capacity), new_chunk_capacity);
+}
+
+template <typename T, size_t max_contiguous_allocation>
+void chunked_vector<T, max_contiguous_allocation>::set_chunk_after_emplace_back(std::pair<chunk_ptr, size_t> x) noexcept {
+    auto& new_last_chunk = x.first;
+    auto& new_chunk_capacity = x.second;
+    auto last_chunk_size = _size % max_chunk_capacity();
+    if (last_chunk_size) {
+        migrate(_chunks.back().get(), _chunks.back().get() + last_chunk_size, new_last_chunk.get());
+        _chunks.back() = std::move(new_last_chunk);
+    } else {
+        _chunks.emplace_back(std::move(new_last_chunk));
     }
+    _capacity = (_chunks.size() - 1) * max_chunk_capacity() + new_chunk_capacity;
```

Comment:
This line even increases my puzzlement of the new change and whether it's correct. The function above that set new_chunk_capacity had a couple of cases where chunks were deliberately not full. But this new calculation assumes they always are.

Reply:
only the first chunk might be partial.  All the others are born full.
We replace the last chunk in place only if `last_chunk_size = _size % max_chunk_capacity()` is not zero - meaning the current chunk is partial, otherwise if the last chunk is full, or empty if we add the first chunk now, we append the newly allocated chunk to the list, and in this case it may or may not be full.

The _capacity calculation assumes that if not _chunks have 1 chunk, then `(_chunks.size() - 1) * max_chunk_capacity()` would be 0, and the new_chunk_capacity is the whole _capacity.
If `_chunks.size() > `, then the assumption is that all chunks so far had to be full for allocating a new chunk, and we just add the new_chunk_capacity to that.  Currently this would also always be equal to `max_chunk_capacity`, but it doesn't have to be.  The code here would allow appending partial chunks if we wish, and growing them incrementally until they fill up.

Reply:
Good, I understand now. I didn't understand the significance of _chunk.size() - 1.